File: Dev/C++/Mascaret/src/VEHA/Kernel/SlotDerived.cpp
#include "VEHA/Kernel/SlotDerived.h"
#include "VEHA/Kernel/Property.h"
#include "VEHA/Kernel/ValueSpecification.h"
#include "VEHA/Kernel/InstanceValue.h"
#include "VEHA/Kernel/InstanceSpecification.h"
#include "VEHA/Kernel/Expression.h"
#include <boost/bind.hpp>

namespace VEHA
{

SlotDerived::SlotDerived(InstanceSpecification * instance, shared_ptr<Property> property)
:Slot(instance,property)
{
	shared_ptr<Property> prop=getDefiningProperty();
	if(prop)
	{
		cerr<<"slot derived:"<<shared_dynamic_cast<Expression>(prop->getDefaultValue())->getExpression() <<" from "<<instance->getName()<<endl;
		_expression=shared_dynamic_cast<Expression>(prop->getDefaultValue()->clone());
		if (!_expression) _expression=shared_ptr<Expression>(new Expression(string("'No Expression'"),prop->getType()));
		_values.clear();
		_expression->_cbOnNeedUpdate.connect(boost::bind(&SlotDerived::notifyMaybeChanged,this));
	}
}

SlotDerived::~SlotDerived()
{
}
void SlotDerived::_update()
{
	if (_expression->getNeedUpdate())
	{
		//Remove old values
		//		map<string,shared_ptr<ValueSpecification> >::iterator it;
		//		for (it=_values.begin();it!=_values.end();it++)
		//			_removeValue(it->first);
		_values.clear();
		//add new values
		shared_ptr<InstanceSpecification> is=shared_dynamic_cast<InstanceSpecification>(_owningInstance->shared_from_this());
		vector<shared_ptr<ValueSpecification> > values
		=_expression->checkAndGetValues(
				_expression->evaluateExpression(is	));
		for(size_t i=values.size();i--;)
			_addValue(values[i]);
	}

}
void SlotDerived::notifyMaybeChanged()
{
	_cbOnChange();
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Kernel/SlotDerived.cs
using System;

using System.Collections.Generic;


namespace Mascaret
{
    public class SlotDerived : Slot
    {

        private Expression expression;
        public Expression Expression
        {
            get { return expression; }
            set { expression = value; }
        }

        public SlotDerived(InstanceSpecification instance, Property property)
            : base(instance, property)
        {
            if (DefiningProperty != null)
            {
                System.Console.WriteLine("Slot deriver : " + ((Expression)DefiningProperty.DefaultValue).ExpressionValue + " from" + instance.name);
                expression = (Expression)DefiningProperty.DefaultValue.clone();
                if (expression == null)
                {
                    expression = new Expression("No expression", DefiningProperty.Type);
                }
            }
        }

        public new bool _canRemoveValue()
        {
            return false;
        }

        public new void addValue(ValueSpecification val)
        {
            System.Console.WriteLine("Can't add values in a derived slot");
        }

        public void notifyMaybeChanged()
        {
            //cbOnChange()
        }

        protected new void _update()
        {
            if (expression.NeedUpdate)
            {
                Values.Clear();
                //List<ValueSpecification> values = expression.checkAndGetValues(expression.evaluateExpression(OwningInstance));
                foreach (ValueSpecification valueSpec in Values.Values)
                {
                    _addValue(valueSpec);
                }
            }
        }

    }
}


